### PR TITLE
rbd-mirror: permit release of local image exclusive lock after force promotion

### PR DESCRIPTION
**Pull request by dillaman**
Fixes: http://tracker.ceph.com/issues/18963
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by dillaman** [comment]
The current code should address the common case when an image is being replayed and then it's force-promoted. I still need to handle the case where the image is bootstrapping up the replayer.

File: src/librbd/ExclusiveLock.cc [hunk 1]
```diff
@@ -40,9 +40,12 @@ bool ExclusiveLock<I>::accept_requests(int *ret_val) const {
   Mutex::Locker locker(ML<I>::m_lock);
 
   bool accept_requests = (!ML<I>::is_state_shutdown() &&
+                          !ML<I>::is_state_pre_releasing() &&
                           ML<I>::is_state_locked() &&
```
Comment (trociny) [inline src/librbd/ExclusiveLock.cc]: @dillaman I can't understand this. `is_state_locked()` ensures the only allowed state is `SATE_LOCKED`, so additionally testing that it is not `STATE_PRE_RELEASING` does not make any difference?

Comment (dillaman) [inline src/librbd/ExclusiveLock.cc]: ```is_state_locked``` covers several states where the actual lock is held on the object -- the ```STATE_PRE_RELEASING``` being one of those states. However, if we are in the process of releasing the lock, we shouldn't be accepting new requests nor accepting new IO.

Comment (dillaman) [inline src/librbd/ExclusiveLock.cc]: Wait -- you are correct. I was confusing that w/ ```is_lock_owner``` which covers several states.

File: src/librbd/ExclusiveLock.cc [hunk 2]
```diff
@@ -52,6 +52,18 @@ bool ExclusiveLock<I>::accept_requests(int *ret_val) const {
 }
 
 template <typename I>
+bool ExclusiveLock<I>::accept_ops() const {
+  Mutex::Locker locker(ML<I>::m_lock);
+  bool accept_ops = (!ML<I>::is_state_shutdown() &&
+                     !ML<I>::is_state_pre_releasing() &&
```
Comment (trociny) [inline src/librbd/ExclusiveLock.cc]: Then this also look unnecessary.

File: src/librbd/io/CopyupRequest.h [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #ifndef CEPH_LIBRBD_IO_COPYUP_REQUEST_H
 #define CEPH_LIBRBD_IO_COPYUP_REQUEST_H
 
-#include "librbd/AsyncOperation.h"
+#include "librbd/io/AsyncOperation.h"
```
Comment (trociny) [inline src/librbd/io/CopyupRequest.h]: maybe move below?

**Comment by dillaman** [comment]
@trociny update pushed

**Pull request merged**